subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/compilerapi/constants/ConstantsAnalysisResult.java

```diff
@@ -16,15 +16,31 @@
 
 package org.gradle.api.internal.tasks.compile.incremental.compilerapi.constants;
 
-
+import java.io.Serializable;
 import java.util.Optional;
 
-public interface ConstantsAnalysisResult {
+public class ConstantsAnalysisResult implements Serializable {
```

Comment:
We should think about changing how this works. It's a bit weird to serialize the builder and pass it from the compiler daemon to the daemon. Instead I think we should eventually (not in this PR) change this so that the mapping is built locally _then_ passed to the daemon. This can be done by implementing our custom `writeObject` which would trigger finalization of the builder. Currently not easily doable because the datastructure that we use _out_ of the builder, the mapping itself, includes classes which are not serializable or not efficiently serialized. This means we should think about a more efficient communication protocol between the compiler daemon and the client.

The advantage is that we would perform the analysis and the mapping building in the compiler daemon itself, which is done (usually) out of process so this leaves the daemon free to do other things in between.

Reply:
I'm not sure what problem that would solve. Have we identified a performance problem here?

Reply:
It's more a logical problem than an performance one. The "builder" should actually live on the compiler worker. There's no need for the builder on the consumer (daemon) side. We do this by convenience only. IF, we want to fix this, it could have the nice side effect of improving performance by delegating the work of building the actual final structure to the compiler worker instead of the daemon, which is free to do other tasks in parallel.